### PR TITLE
[Core][Model] Add simple_model_runner and a new model XLMRobertaForSequenceClassification through multimodal interface

### LINKED ISSUE #6424: [Feature]: Return softmax of attention layer.
### 🚀 The feature, motivation and pitch

# Feature
I'm working on returning the softmax of the LLM's attention layer. As far as I know, only transformers support this [feature](https://huggingface.co/docs/transformers/v4.42.0/en/internal/generation_utils#generate-outputs). But transforms is not efficient enough. I saw "return_softmax" [vllm_flash_attn](https://github.com/vllm-project/flash-attention/blob/main/vllm_flash_attn/flash_attn_interface.py#L62), but seems it can't be used. Would you be able to add this feature?

# Motivation
Softmax of attention is important evidence for humans to understand how LLM works. Humans can determine which input token is important for the inference of the output token. And more features can be developed based on this feature such as visualization of LLM evidence.

# Pitch
For this RFC in particular, we propose the following changes:
- Add a new function llm.forward(). This function receives requests to do prefill and return softmax.
- Modify prefill kernels to support output softmax.

### Alternatives

_No response_

### Additional context

_No response_

### LINKED ISSUE #6789: [Usage]: can I use it with classification model (e.g. GemmaForSequenceClassification) ?
### Your current environment

```text
The output of `python collect_env.py`
```

### How would you like to use vllm

can I use it with classification model (e.g. GemmaForSequenceClassification) ?

### LINKED ISSUE #8022: [New Model]: Want to support BAAI/bge-reranker-v2-m3 model
### The model to consider.

https://huggingface.co/BAAI/bge-reranker-v2-m3

### The closest model vllm already supports.

_No response_

### What's your difficulty of supporting the model you want?

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by AllenDou**
This PR, 
1. Add a new model XLMRobertaForSequenceClassification (for RAG scenario) [https://github.com/huggingface/transformers/blob/v4.42.3/src/transformers/models/xlm_roberta/modeling_xlm_roberta.py](url)
which processes input data through a multimodal interface like the following
```
outputs = llm.process([{
    "prompt": prompt,
    "multi_modal_data": {
        "xlmroberta": inputs,
    }
}])
```
2. Introduce `class ModelMode [DECODER, ENCODER, ENCODER_DECODER, EMBEDDING, SIMPLE]`
3. Add a new model runner(simple_model_runner.py), I didn't use model_runner.py or embedding_model_runner.py because they have procedures like logits(), sample(), and pooling() that we don't need. Maybe we could find a way to create a more general model runner to support 'any' model
4. Add SimpleSequenceGroupOutput, SimpleRequestOutput, and SimpleOutput for general output
5. Tensor-parallel is enabled, but no benefits have been achieved so far
6. Test case/example added

I have two goals:
1. To find a way to enable vllm to integrate 'any' model
2. To replace these models' layers with vllm/layers/* as well as TP/PP to achieve performance improvement

CLOSE #6424
CLOSE #6789
CLOSE #8022

### CODE REVIEW

**Comment by DarkLight1337** [comment]
If I'm understanding this PR correctly, you are basically using the multi-modal interface to pass data directly to the model (in this case the input IDs and attention mask). Are you working towards making vLLM function out-of-the-box with generic HuggingFace models?

File: examples/flagembedding.py
Comment (DarkLight1337) [inline examples/flagembedding.py:6]: I don't quite see the value of this example in vLLM. Are you planning to update it to use vLLM components?

File: tests/conftest.py
Comment (DarkLight1337) [inline tests/conftest.py:424]: Imo the naming is a bit misleading since you're returning the logits of the classification head, not the encoded input directly.

File: vllm/entrypoints/llm.py
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:426]: Same here, it's not really "encoding" the input.

**Review by DarkLight1337 — COMMENTED** [review]
Left some initial comments.

**Comment by AllenDou** [comment]
> If I'm understanding this PR correctly, you are basically using the multi-modal interface to pass data directly to the model (in this case the input IDs and attention mask). Are you working towards making vLLM function out-of-the-box with generic HuggingFace models?

Yes, you are right. In fact, I have two goals:
1. To find a way to enable vllm to integrate 'any' model.
2. To replace these models' layers with vllm/layers/* as well as TP/PP to achieve performance improvement.

After I replaced XLMRobertaForSequenceClassification's query/key/value linear with QKVParallelLinear, I saw a 15% performance improvement.

File: examples/flagembedding.py
Comment (AllenDou) [reply examples/flagembedding.py:6]: In RAG scenario, user usually use FlagEmbedding package, I left it as an example, but could be removed.

File: tests/conftest.py
Comment (AllenDou) [reply tests/conftest.py:424]: Yes, you are right, I use encode_* because I did't find a proper func name for this general 'any' model scenario, maybe someone could give me some suggestion.

Comment (DarkLight1337) [reply tests/conftest.py:424]: Hmm I think since this is so general, you can extract the logits inside your specific test rather than in `conftest.py`. I'm tempted to just call it `forward` since all it does is pass the inputs directly to the model.

Comment (AllenDou) [reply tests/conftest.py:424]: Good idea, now it's late, I'll refactor it tomorrow, thank for your advice :)

File: vllm/entrypoints/llm.py
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py]: We should still check `simple_mode` here imo. Perhaps we should make the mode an enum to avoid complex if-else cases as we add more modes..

File: vllm/entrypoints/llm.py
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py]: Should update the docstring here.

File: vllm/entrypoints/llm.py
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:491]: To simplify the code, it may be a good idea to drop the old API for this new method.

Comment (AllenDou) [reply vllm/entrypoints/llm.py:491]: Sorry, I didn't get your point, the 'old API' mean?  prompt_token_ids? prompts? these signature?

File: vllm/entrypoints/llm.py
Comment (AllenDou) [reply vllm/entrypoints/llm.py]: Good idea, Let me try.

File: vllm/entrypoints/llm.py
Comment (DarkLight1337) [reply vllm/entrypoints/llm.py:491]: Yeah, that's what I mean. We can have this function only accept `inputs`.

File: vllm/entrypoints/llm.py
Comment (AllenDou) [reply vllm/entrypoints/llm.py]: I created a class ModelMode to indicate the model mode [ENCODER, DECODER, EMBEDDING, SIMPLE]. Please take a look and give me some suggestions :)

File: vllm/config.py
Comment (DarkLight1337) [inline vllm/config.py:193]: Let's combine these two functions together and move the logic of determining model mode to model registry. (May have to move `ModelMode` as well to avoid circular imports)

File: vllm/engine/llm_engine.py
Comment (DarkLight1337) [inline vllm/engine/llm_engine.py:265]: Perhaps we should let each mode declare whether to set up KV cache instead of handling each case inside `LLMEngine`.

File: vllm/core/scheduler.py
Comment (DarkLight1337) [inline vllm/core/scheduler.py:285]: Perhaps we should let each mode override the `BlockSpaceManagerImpl` instead of handling each case inside `Scheduler`.

File: vllm/engine/llm_engine.py
Comment (DarkLight1337) [inline vllm/engine/llm_engine.py:770]: I wonder whether we could also factor out this logic into the individual `ModelMode`s....

File: vllm/entrypoints/llm.py
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:290]: I think whitelisting the model modes would be easier to maintain instead of blacklisting them.

File: vllm/entrypoints/llm.py
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:437]: `is not` check should work with enums.

File: vllm/entrypoints/openai/serving_embedding.py
Comment (DarkLight1337) [inline vllm/entrypoints/openai/serving_embedding.py:140]: `is not` check should work with enums.

File: vllm/worker/worker.py
Comment (DarkLight1337) [inline vllm/worker/worker.py:89]: Perhaps we should let each mode override the `ModelRunnerClass` instead of handling each case inside `Worker`.

File: vllm/entrypoints/llm.py
Comment (DarkLight1337) [reply vllm/entrypoints/llm.py]: Added a new round of comments!

File: vllm/worker/worker.py
Comment (AllenDou) [reply vllm/worker/worker.py:89]: I made some changes as you can see
But I didn't quite understand your point about "override BlockSpaceManagerImpl ModelRunnerClass". Do you mean to override the entire BlockManager/ModelRunner class? That seems more complicated than the current implementation.

Comment (DarkLight1337) [reply vllm/worker/worker.py:89]: Here, by "override" I just mean set the `ModelRunnerClass` to use. We can add static methods to `ModelMode` (e.g. `get_block_space_manager_impl` and `get_model_runner_cls`) so we can keep the switch/case code all in one place.

Comment (AllenDou) [reply vllm/worker/worker.py]: Please take a look class ModelMode, some 'circular import' problems occur, but solved.

File: vllm/model_executor/models/__init__.py
Comment (DarkLight1337) [inline vllm/model_executor/models/__init__.py:186]:
```suggestion
    def get_model_mode(architectures: List[str]) -> ModelMode:
```

File: vllm/model_executor/models/__init__.py
Comment (DarkLight1337) [inline vllm/model_executor/models/__init__.py:202]:
```suggestion
    def need_initialize_kv_caches(model_mode: ModelMode) -> bool:
        if model_mode == ModelMode.EMBEDDING:
            return False
        if model_mode == ModelMode.SIMPLE:
            return False
        return True
```

File: vllm/model_executor/models/__init__.py
Comment (DarkLight1337) [inline vllm/model_executor/models/__init__.py:98]:
```suggestion
    def get_model_runner_cls(model_mode: "ModelMode"):
```

File: vllm/model_executor/models/__init__.py
Comment (DarkLight1337) [inline vllm/model_executor/models/__init__.py:109]:
```suggestion
    def get_block_space_manager_impl(use_v2_block_manager: bool, model_mode: "ModelMode"):
```

**Review by DarkLight1337 — COMMENTED** [review]
Added some suggestions to improve the type annotations.

File: vllm/config.py
Comment (DarkLight1337) [inline vllm/config.py:156]: Maybe can remove `_verify_model_mode` and just set it here.

File: vllm/config.py
Comment (DarkLight1337) [inline vllm/config.py:787]: Can also factor this out to `ModelMode`. (Return `_EMBEDDING_MODEL_MAX_NUM_BATCHED_TOKENS` etc. inside `ModelMode`, but keep the `max(...)` logic here.)

Comment (AllenDou) [reply vllm/config.py]: Please take a look :)

File: vllm/config.py
Comment (AllenDou) [reply vllm/config.py]: done.

File: vllm/config.py
Comment (DarkLight1337) [inline vllm/config.py:776]: How about we return `None` for the other model modes (so we don't need the `elif` statement for this)? Afterwards we only apply `max` if the model mode returns a valid integer.

Comment (AllenDou) [reply vllm/config.py:776]: Only keep "if enable_chunked_prefill" "else"?   remove "elif:"?

Comment (DarkLight1337) [reply vllm/config.py:776]: Yeah. We can apply the `max` part inside the `else` block

Comment (AllenDou) [reply vllm/config.py:776]: Sorry, didn't get your point.

File: vllm/model_executor/models/__init__.py
Comment (DarkLight1337) [inline vllm/model_executor/models/__init__.py:136]:
```suggestion
    def get_model_max_num_batched_tokens(model_mode: "ModelMode") -> Optional[int]:
```

File: vllm/core/interfaces.py
Comment (DarkLight1337) [inline vllm/core/interfaces.py:47]: Can we remove the "embedding" and "simple" versions and instead return the class directly inside `ModelMode`? (Only keep "v1" and "v2" since they are actually versions)

Comment (DarkLight1337) [reply vllm/core/interfaces.py:47]: Edit: Actually I think this function isn't used anymore?

File: vllm/worker/worker.py
Comment (DarkLight1337) [inline vllm/worker/worker.py:89]: I think we can simplify this to a one-liner

```py
ModelRunnerClass = model_runner_cls or ModelMode.get_model_runner_cls(self.model_config.model_mode)
```

Comment (AllenDou) [reply vllm/worker/worker.py:89]: Yes, get_block_space_manager_class() is not used anymore, replaced by get_block_space_manager_impl()

File: vllm/core/interfaces.py
Comment (AllenDou) [reply vllm/core/interfaces.py]: removed.

**Review by DarkLight1337 — APPROVED** [review]
The config and multi-modal parts LGTM. I assume the model is implemented correctly since the tests pass.

However, since I'm not involved with the internals of model executor, block manager and workers, I'll leave it to @robertgshaw2-neuralmagic to review those. He will also see how to integrate this into the existing work for encoder-decoder support.

**Comment by robertgshaw2-neuralmagic** [comment]
Im going to take a look at this over the weekend. Thanks @AllenDou!

**Comment by AllenDou** [comment]
mark https://github.com/vllm-project/vllm/issues/6424

**Comment by AllenDou** [comment]
Hello @robertgshaw2-neuralmagic , just a friendly reminder to review this PR when you get a chance,
We want to enable more models through this feature,  thank a lot!

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks @AllenDou ! this is on my list

**Comment by AllenDou** [comment]
mark https://github.com/vllm-project/vllm/issues/6789

**Comment by AllenDou** [comment]
mark https://github.com/vllm-project/vllm/issues/1187 https://github.com/vllm-project/vllm/issues/205

**Comment by DarkLight1337** [comment]
I think @robertgshaw2-neuralmagic has been very busy lately... Are you still interested in implementing this? Let's split the PR into smaller parts that are more manageable so that @robertgshaw2-neuralmagic doesn't have to review everything at once.

(By the way, please take into account the changes from #4942.)

**Comment by AllenDou** [comment]
> I think @robertgshaw2-neuralmagic has been very busy lately... Are you still interested in implementing this? Let's split the PR into smaller parts that are more manageable so that @robertgshaw2-neuralmagic doesn't have to review everything at once.
> 
> (By the way, please take into account the changes from #4942.)

Of course, after reviewing https://github.com/vllm-project/vllm/pull/4942, I realized that nm folks have their own methods for supporting 'anymodel'. So, I think maybe I should close this PR. If this feature is needed in the future, I can reopen this PR at any time.

**Comment by robertgshaw2-neuralmagic** [comment]
Hey @AllenDou Im really sorry for the delay here, have had no bandwidth

#4942 is only for encoder-decoder, which is separate from the implementation here, so I think we should be able to include both

**Comment by AllenDou** [comment]
> Hey @AllenDou Im really sorry for the delay here, have had no bandwidth over the past month
> 
> #4942 is only for encoder-decoder, which is separate from the feature here, so I do not think this work is a duplicate
> 
> I’m happy to pick this back up if you want to reopen the PR

Great, I'll resolve the conflicts and reopen the PR, I hope this won't take up too much of your time :)

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks so much, I really appreciate your patience. Resolving our performance issues has just become a huge priority

**Comment by AllenDou** [comment]
wwww

**Comment by AllenDou** [comment]
Hi @robertgshaw2-neuralmagic , I have rebased and reopened this PR. Please take a look at it. The offline_inference_xlmroberta_awq.py file is a quantized version I've been working on, but I'm encountering some technical difficulties that are proving challenging to overcome. So, this file could be removed if necessary. cc @DarkLight1337

**Comment by AllenDou** [comment]
offline_inference_xlmroberta_awq.py is deleted, as after hacking autoawq for xlmroberta model, I see no gain of performance benifit under vllm serving. Trying to fp8.

**Comment by prashil1996** [comment]
@AllenDou @robertgshaw2-neuralmagic @DarkLight1337 
Is there any timeline on when this could be delivered?

**Comment by AllenDou** [comment]
> @AllenDou @robertgshaw2-neuralmagic @DarkLight1337 Is there any timeline on when this could be delivered?

Maybe we should wait until @robertgshaw2-neuralmagic gets a chance to review this PR.

**Comment by DarkLight1337** [comment]
A quick heads-up that the new locations of the model tests have been adjusted in #7820, so please merge from main.

**Comment by DarkLight1337** [comment]
Also @robertgshaw2-neuralmagic do you have any timeframe on when you will be available to review this PR?

**Comment by fan-niu** [comment]
@AllenDou Hello, thank you for your contribution. I use concurrency to request services, and there will be a series of error reports, including the following errors:
` File "/share5/projects/llm/finetune-accelerate/inference/tensorrtllm-experiment/sglang/code/vllm_test/vllm_sqclassification/vllm/vllm/model_executor/models/xlmroberta.py", line 721, in forward
    batch_size, seq_length = input_shape
ValueError: not enough values to unpack (expected 2, got 1)`

Can you please help to optimize async requests? Thanks

Note: 
I use  http://127.0.0.1:8081/v1/completions, concurrency = 2,
payload:
{
        "prompt": "what time is it",
        "model":MODEL_NAME,
        "stop": "<\s>",
        "max_tokens":128,
        "stream": False,
        "presence_penalty": 0,
        "frequency_penalty": 0,
        "temperature": 0.0,
        "top_p": 1,
    }

**Comment by AllenDou** [comment]
> @AllenDou Hello, thank you for your contribution. I use concurrency to request services, and there will be a series of error reports, including the following errors: ` File "/share5/projects/llm/finetune-accelerate/inference/tensorrtllm-experiment/sglang/code/vllm_test/vllm_sqclassification/vllm/vllm/model_executor/models/xlmroberta.py", line 721, in forward batch_size, seq_length = input_shape ValueError: not enough values to unpack (expected 2, got 1)`
> 
> Can you please help to optimize async requests? Thanks
> 
> Note: I use http://127.0.0.1:8081/v1/completions, concurrency = 2, payload: { "prompt": "what time is it", "model":MODEL_NAME, "stop": "<\s>", "max_tokens":128, "stream": False, "presence_penalty": 0, "frequency_penalty": 0, "temperature": 0.0, "top_p": 1, }

This PR currently does not support frontend access through the http protocol. By the way, the XLM-Roberta model compares the similarity of two strings. Therefore, you need to pass a tuple containing two strings (string, string) as input. Please refer to examples/offline_inference_xlmroberta.py for more details.

**Comment by fan-niu** [comment]
> [RFC]: Support encode only models by Workflow Defined Engine #8453

Thanks for your reply, can you tell me how to add support for http requests on your branch? Thank you, or can you add this function if it is convenient for you?

**Comment by wangyuanxiong-hub** [comment]
> @AllenDou Thanks for your reply, can you tell me how to add support for http requests on your branch? Thank you, or can you add this function if it is convenient for you? My final goal is to need llamaForSequenceClassification based on http request, I have added llamaForSequenceClassification to llama.py with reference to your branch, and implemented the forward function, but currently I cannot make the correct request through http.

Is it accessible via http now?

**Comment by wangyuanxiong-hub** [comment]
The test results were not as expected, and the input and output methods were very confusing.

**Comment by AllenDou** [comment]
> > @AllenDou Thanks for your reply, can you tell me how to add support for http requests on your branch? Thank you, or can you add this function if it is convenient for you? My final goal is to need llamaForSequenceClassification based on http request, I have added llamaForSequenceClassification to llama.py with reference to your branch, and implemented the forward function, but currently I cannot make the correct request through http.
> 
> Is it accessible via http now?

I apologize for the delayed response. I've been working on cute&ptx recently, therefore, this PR will not be updated for now :(

**Comment by DarkLight1337** [comment]
Closing as superseded by #10400. Sorry your PR didn't make it!

**Pull request closed without merging**
